src/category_theory/monoidal/category.lean

```diff
@@ -0,0 +1,405 @@
+by { rw [←interchange, ←interchange], simp }
+
+lemma left_unitor_inv_naturality {X X' : C} (f : X ⟶ X') :
+  f ≫ (left_unitor X').inv = (left_unitor X).inv ≫ (𝟙 _ ⊗ f) :=
+begin
+  apply (cancel_mono (left_unitor X').hom).1,
+  simp only [assoc, comp_id, iso.inv_hom_id],
+  conv_rhs { congr, skip, rw [left_unitor_naturality] },
+  conv_rhs { rw [←category.assoc], congr, rw [iso.inv_hom_id] },
+  conv_rhs { rw [category.id_comp] }
+end
+
+lemma right_unitor_inv_naturality {X X' : C} (f : X ⟶ X') :
+  f ≫ (right_unitor X').inv = (right_unitor X).inv ≫ (f ⊗ 𝟙 _) :=
+begin
+  apply (cancel_mono (right_unitor X').hom).1,
+  simp only [assoc, comp_id, iso.inv_hom_id],
+  conv_rhs { congr, skip, rw [right_unitor_naturality] },
+  conv_rhs { rw [←category.assoc], congr, rw [iso.inv_hom_id] },
+  conv_rhs { rw [category.id_comp] }
+end
+
+@[simp] lemma tensor_left_iff
+  {X Y : C} (f g : X ⟶ Y) :
+  ((𝟙 (tensor_unit C)) ⊗ f = (𝟙 (tensor_unit C)) ⊗ g) ↔ (f = g) :=
+begin
+  split,
+  { intro h,
+    have h' := congr_arg (λ k, (left_unitor _).inv ≫ k) h,
+    dsimp at h',
+    rw ←left_unitor_inv_naturality at h',
+    rw ←left_unitor_inv_naturality at h',
+    exact (cancel_mono _).1 h', },
+  { intro h, subst h, }
+end
+
+@[simp] lemma tensor_right_iff
+  {X Y : C} (f g : X ⟶ Y) :
+  (f ⊗ (𝟙 (tensor_unit C)) = g ⊗ (𝟙 (tensor_unit C))) ↔ (f = g) :=
+begin
+  split,
+  { intro h,
+    have h' := congr_arg (λ k, (right_unitor _).inv ≫ k) h,
+    dsimp at h',
+    rw ←right_unitor_inv_naturality at h',
+    rw ←right_unitor_inv_naturality at h',
+    exact (cancel_mono _).1 h' },
+  { intro h, subst h, }
+end
+
+-- We now prove:
+--   ((associator (tensor_unit C) X Y).hom) ≫
+--     ((left_unitor (X ⊗ Y)).hom)
+--   = ((left_unitor X).hom ⊗ (𝟙 Y))
+-- (and the corresponding fact for right unitors)
+-- following the proof on nLab.
+
+lemma left_unitor_product_aux_perimeter (X Y : C) :
+    ((associator (tensor_unit C) (tensor_unit C) X).hom ⊗ (𝟙 Y)) ≫
+    (associator (tensor_unit C) ((tensor_unit C) ⊗ X) Y).hom ≫
+    ((𝟙 (tensor_unit C)) ⊗ (associator (tensor_unit C) X Y).hom) ≫
+    ((𝟙 (tensor_unit C)) ⊗ (left_unitor (X ⊗ Y)).hom)
+  = (((right_unitor (tensor_unit C)).hom ⊗ (𝟙 X)) ⊗ (𝟙 Y)) ≫
+    (associator (tensor_unit C) X Y).hom :=
+begin
+  conv_lhs { congr, skip, rw [←category.assoc] },
+  conv_lhs { rw [←category.assoc], congr, rw [monoidal_category.pentagon] },
+  conv_rhs { rw [associator_naturality], congr, skip, congr, skip, rw [tensor_map_id] },
+  conv_rhs { congr, skip, rw [←monoidal_category.triangle] },
+  conv_rhs { rw [←category.assoc] }
+end
+
+lemma left_unitor_product_aux_triangle (X Y : C) :
+    ((associator (tensor_unit C) (tensor_unit C) X).hom ⊗ (𝟙 Y)) ≫
+    (((𝟙 (tensor_unit C)) ⊗ (left_unitor X).hom) ⊗ (𝟙 Y))
+  = ((right_unitor (tensor_unit C)).hom ⊗ (𝟙 X)) ⊗ (𝟙 Y) :=
+begin
+  conv_lhs { rw [←interchange_left_identity] },
+  conv_rhs { congr, rw [←monoidal_category.triangle] }
```

Comment:
I'm confused by `conv_rhs { congr, ... }`. Doesn't `conv_rhs` narrow to the right-hand side of the equation, and how does `congr` make sense then?

In any case it looks like the proof could be simplified, if I'm not missing something.

Reply:
Sorry, yes, all these `rewrite` proofs are just straight out of `rewrite_search`, which is slowly lumbering towards PRs. I'll clean them up now.